### PR TITLE
pkg/reloader: improve detection of directory changes

### PR DESCRIPTION
**Pull request by simonpasquier**
* [X] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

When watching for changes in directories, the reloader used to rely only
on the watch interval and not on the inotify events. This commit implements
a more efficient detection of changes for watched directories.

The change also adds a new `DelayInterval` option that allows to delay
the config reload after no additional event are received.

Finally a new metric,
`thanos_sidecar_reloader_config_apply_operations_total`, is added and
`thanos_sidecar_reloader_config_apply_errors_total` has been renamed
to `thanos_sidecar_reloader_config_apply_operations_failed_total` for
consistency.

## Verification

I've modified the unit test to validate that the reloader reacts on file changes (without the periodic watch). I've added a unit test to validate that the reloader's periodic watch detects a change in case it hasn't been caught by fsnotify.

I've also included this PR in https://github.com/prometheus-operator/prometheus-operator/pull/3457 which is the change that triggered this PR.

### CODE REVIEW

**Comment by kakkoyun** [comment]
@simonpasquier will this have any effect on falky reloader tests https://github.com/thanos-io/thanos/issues/2622? 

Also e2e tests are failing.

**Comment by simonpasquier** [comment]
Kemal, thanks for pointing the issue, I'll look into it.

File: pkg/reloader/reloader.go [hunk 1]
```diff
@@ -207,42 +201,46 @@ func (r *Reloader) Watch(ctx context.Context) error {
...
+	}()
 
-	r.watches.Set(float64(len(watchables)))
 	level.Info(r.logger).Log(
 		"msg", "started watching config file and directories for changes",
 		"cfg", r.cfgFile,
 		"out", r.cfgOutputFile,
 		"dirs", strings.Join(r.watchedDirs, ","))
 
+	applyCtx, cancel := context.WithTimeout(ctx, r.watchInterval)
```
Comment (kakkoyun) [inline pkg/reloader/reloader.go:224]: nit: Renaming the `cancel` to `applyCancel` could make it easier to track.
```suggestion
	applyCtx, applyCancel := context.WithTimeout(ctx, r.watchInterval)
```

File: pkg/reloader/reloader.go [hunk 2]
```diff
@@ -434,3 +428,149 @@ func expandEnv(b []byte) (r []byte, err error) {
...
+			prometheus.CounterOpts{
+				Name: "reloader_watch_errors_total",
+				Help: "Total number of errors received by the reloader from the watcher.",
+			},
+		),
+	}
+}
+
+func (w *watcher) addPath(name string) error {
+	if w.w == nil {
```
Comment (kakkoyun) [inline pkg/reloader/reloader.go:474]: nit: Does it ever get called concurrently? If it's the case we could use `sync.Once` for initialization.

File: pkg/reloader/reloader.go [hunk 3]
```diff
@@ -434,3 +428,149 @@ func expandEnv(b []byte) (r []byte, err error) {
...
+func (w *watcher) addFile(name string) error {
+	w.watchedDirs[filepath.Dir(name)] = struct{}{}
+	return w.addPath(name)
+}
+
+func (w *watcher) run(ctx context.Context) {
+	defer runutil.CloseWithLogOnErr(w.logger, w.w, "config watcher close")
+
+	var wg sync.WaitGroup
+	notify := make(chan struct{})
```
Comment (kakkoyun) [inline pkg/reloader/reloader.go:506]: nit
```suggestion
	var (
		wg sync.WaitGroup
		notify = make(chan struct{})
	)
```

**Review by kakkoyun — APPROVED** [review]
Amazing work. LGTM AFAICT.

**Comment by kakkoyun** [comment]
Let's have another pair of eyes before we merge this.  cc @thanos-io/thanos-maintainers

File: pkg/reloader/reloader.go (hunk 2 above)
Comment (simonpasquier) [reply pkg/reloader/reloader.go:474]: No, `addPath` and `addDir` are called sequentially in `*Reloader.Watch()`.

File: pkg/reloader/reloader.go [hunk 4]
```diff
@@ -149,56 +154,45 @@ func New(logger log.Logger, reg prometheus.Registerer, o *Options) *Reloader {
...
+// Because some edge cases might be missing, the reloader also relies on the
+// watch interval.
 func (r *Reloader) Watch(ctx context.Context) error {
+	if r.cfgFile == "" && len(r.watchedDirs) == 0 {
+		level.Info(r.logger).Log("msg", "nothing to be watched")
+		<-ctx.Done()
+		return nil
+	}
+
 	watcher, err := fsnotify.NewWatcher()
```
Comment (simonpasquier) [inline pkg/reloader/reloader.go:188]: oops this needs to be removed.

**Comment by simonpasquier** [comment]
prometheus-operator/prometheus-operator#3457 is the prometheus-operator PR including this change.

I'm not sure that this PR will help to solve https://github.com/thanos-io/thanos/issues/2622. @kakkoyun do you have a recent CI failure for this? I've run the unit tests several times (>100) on my machine with `-race -cpu 2` and all passed (both on `master` and with the PR branch).

**Comment by kakkoyun** [comment]
Thanks a lot for addressing issues @simonpasquier 💜 The flakiness happens on a single core and relatively slow CPU setups (such as CI). It's ok, it's not our priority for this PR. Let's see if it helps after merging this.

**Review by brancz — APPROVED** [review]
Nice!

**Pull request merged**